handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslClientContext.java

```diff
@@ -100,18 +99,23 @@ static OpenSslSessionContext newSessionContext(ReferenceCountedOpenSslContext th
                     setKeyMaterial(ctx, keyCertChain, key, keyPassword);
                 }
             } else {
+                OpenSslKeyMaterialProvider keyMaterialProvider = null;
+
                 // javadocs state that keyManagerFactory has precedent over keyCertChain
                 if (keyManagerFactory == null && keyCertChain != null) {
-                    keyManagerFactory = buildKeyManagerFactory(
-                            keyCertChain, key, keyPassword, keyManagerFactory);
+                    keyMaterialProvider = new OpenSslStaticX509KeyManagerFactory(buildKeyManagerFactory(
+                            keyCertChain, key, keyPassword, null), keyPassword).provider();
+                } else if (keyManagerFactory != null) {
+                    if (keyManagerFactory instanceof OpenSslStaticX509KeyManagerFactory) {
+                        keyMaterialProvider = ((OpenSslStaticX509KeyManagerFactory) keyManagerFactory).provider();
+                    } else {
+                        X509KeyManager keyManager = chooseX509KeyManager(keyManagerFactory.getKeyManagers());
+                        keyMaterialProvider = new OpenSslKeyMaterialProvider(keyManager, keyPassword);
+                    }
                 }
 
-                if (keyManagerFactory != null) {
-                    X509KeyManager keyManager = chooseX509KeyManager(keyManagerFactory.getKeyManagers());
-                    OpenSslKeyMaterialManager materialManager = useExtendedKeyManager(keyManager) ?
-                            new OpenSslExtendedKeyMaterialManager(
-                                    (X509ExtendedKeyManager) keyManager, keyPassword) :
-                            new OpenSslKeyMaterialManager(keyManager, keyPassword);
+                if (keyMaterialProvider != null) {
+                    OpenSslKeyMaterialManager materialManager = new OpenSslKeyMaterialManager(keyMaterialProvider);
```

Comment:
`ReferenceCountedOpenSslClientContext#keyMaterialManager()` always returns `null`, but this method is called in `ReferenceCountedOpenSslContext#destroy()` to free up resources. Should we use a similar strategy for initialization as `ReferenceCountedOpenSslServerContext` to make sure this doesn't leak?

Reply:
Good catch, let me fix it.

Reply:
fixed